File: pydockenv/commands/dependency.py
import click

from pydockenv.executor import Executor


def _build_install_args(packages, requirements_file):
    if not isinstance(packages, list):
        packages = [packages]

    args = ['pip', 'install']
    if requirements_file:
        args.extend(['-r', requirements_file])
    else:
        args.extend(packages)

    return args


def install(packages, requirements_file):
    return Executor.execute(
        *_build_install_args(packages, requirements_file))


def install_for_container(container, packages, requirements_file):
    return Executor.execute_for_container(
        container, *_build_install_args(packages, requirements_file),
        bypass_check=True)


def uninstall(package, yes):
    click.echo('Running...')
    args = ['pip', 'uninstall']
    if yes:
        args.append('-y')

    args.append(package)
    try:
        Executor.execute(*args)
    finally:
        click.echo('Exited!')


def list_packages():
    click.echo('Running...')
    try:
        Executor.execute('pip', 'freeze')
    finally:
        click.echo('Exited!')

File: internal/dependency/dependency.go
package dependency

import (
	"fmt"

	"github.com/docker/docker/api/types"
	"github.com/se7entyse7en/pydockenv/internal/executor"
)

type Requirements struct {
	FileName string
	Packages
}

type Packages struct {
	Dependencies    map[string]string
	RawDependencies []string
}

func Install(requirements *Requirements) error {
	cmd := buildInstallCmd(requirements)
	err := executor.Execute(cmd, &executor.ExecOptions{})
	if err != nil {
		return fmt.Errorf("cannot install requirements in container: %w", err)
	}

	return nil
}

func InstallForContainer(container types.ContainerJSON, requirements *Requirements) error {
	cmd := buildInstallCmd(requirements)
	err := executor.ExecuteForContainer(container, cmd,
		&executor.ExecOptions{ByPassCheck: true})
	if err != nil {
		return fmt.Errorf("cannot install requirements in container: %w", err)
	}

	return nil
}

func Uninstall(packages *Packages, yes bool) error {
	cmd := []string{"pip", "uninstall"}
	cmd = append(cmd, parsePackages(packages)...)
	if yes {
		cmd = append(cmd, "-y")
	}

	err := executor.Execute(cmd, &executor.ExecOptions{})
	if err != nil {
		return fmt.Errorf("cannot uninstall requirements in container: %w", err)
	}

	return nil
}

func ListPackage() error {
	err := executor.Execute([]string{"pip", "freeze"}, &executor.ExecOptions{})
	if err != nil {
		return fmt.Errorf("cannot list packages in container: %w", err)
	}

	return nil
}

func buildInstallCmd(requirements *Requirements) []string {
	cmd := []string{"pip", "install"}
	if requirements.FileName != "" {
		cmd = append(cmd, "-r", requirements.FileName)
	} else {
		cmd = append(cmd, parsePackages(&requirements.Packages)...)
	}

	return cmd
}

func parsePackages(packages *Packages) []string {
	if len(packages.RawDependencies) > 0 {
		return packages.RawDependencies
	}

	var parsedPackages []string
	for p, v := range packages.Dependencies {
		parsedPackages = append(parsedPackages, fmt.Sprintf("%s%s", p, v))
	}

	return parsedPackages
}
